### PR TITLE
fix: preserve Slack draft previews after send

### PR DESCRIPTION
**Pull request by elie222**
# User description
Keeps Slack draft notifications readable after a draft is sent and fixes preview formatting for mixed text and HTML content.

- keep the sent Slack draft card summary and preview instead of collapsing it to a one-line receipt
- normalize lightweight HTML links in Slack draft previews so referral links render correctly
- add focused tests for Slack preview formatting and post-send card behavior

---

# Generated description

Below is a concise technical summary of the changes proposed in this PR:
Preserve Slack draft cards after sending by revising the rule-notifications handler to build a handled notification card that keeps the final summary/preview and reports the sent status instead of collapsing the message. Normalize Slack preview formatting via <code>richTextToSlackMrkdwn</code> so draft and source text that contain lightweight HTML produce readable links and line breaks when rendered in the notification card.
<table><tr><th>Topic</th><th>Details</th><tr><td><a href=https://baz.co/changes/elie222/inbox-zero/2088?tool=ast&topic=Slack+draft+UI>Slack draft UI</a>
        </td><td>Maintain Slack draft notification cards after sending by building a handled notification card that retains the summary, rich preview, and sent status, backed by focused tests on post-send behavior.<details><summary>Modified files (2)</summary><ul><li>apps/web/utils/messaging/rule-notifications.test.ts</li>
<li>apps/web/utils/messaging/rule-notifications.ts</li></ul></details><details><summary>Latest Contributors(1)</summary><table><tr><th>User</th><th>Commit</th><th>Date</th></tr><tr><td>elie222</td><td>assistant: add messagi...</td><td>March 29, 2026</td></tr></table></details></td></tr>
<tr><td><a href=https://baz.co/changes/elie222/inbox-zero/2088?tool=ast&topic=Rich+text+format>Rich text format</a>
        </td><td>Normalize Slack rich-text conversions so HTML links, structure, and edge cases (entities, angle brackets, placeholders) render correctly in previews before they reach Slack.<details><summary>Modified files (2)</summary><ul><li>apps/web/utils/messaging/providers/slack/format.test.ts</li>
<li>apps/web/utils/messaging/providers/slack/format.ts</li></ul></details><details><summary>Latest Contributors(1)</summary><table><tr><th>User</th><th>Commit</th><th>Date</th></tr><tr><td>elie222</td><td>Messaging: migrate to ...</td><td>March 02, 2026</td></tr></table></details></td></tr></table>
This pull request is reviewed by Baz. Review like a pro on <a href=https://baz.co/changes/elie222/inbox-zero/2088?tool=ast>(Baz)</a>.

### CODE REVIEW

File: apps/web/utils/messaging/rule-notifications.ts [hunk 1]
```diff
@@ -873,11 +900,32 @@ function buildTerminalCard({
...
+  content,
+  status,
+}: {
+  content: NotificationContent;
+  status: string;
+}): CardElement {
+  const children: CardChild[] = [CardText(content.summary)];
+
+  if (content.preview) {
+    children.push(CardText(content.preview));
```
[the remark this answers is not in the document]
Comment (elie222) [reply apps/web/utils/messaging/rule-notifications.ts]: Addressed in d47298a. I extracted the shared summary and preview assembly into `buildNotificationCardBody(content)` and reused it in both card builders to remove the duplication.

File: apps/web/utils/messaging/providers/slack/format.ts [hunk 2]
```diff
@@ -24,3 +24,25 @@ export function markdownToSlackMrkdwn(text: string): string {
...
+/**
+ * Normalizes lightweight HTML that can leak into draft previews before
+ * converting the result to Slack mrkdwn.
+ */
+export function richTextToSlackMrkdwn(text: string): string {
+  return markdownToSlackMrkdwn(
+    text
+      .replace(/<br\s*\/?>/gi, "\n")
+      .replace(/<\/(p|div|blockquote|section|article)>/gi, "\n")
+      .replace(/<li\b[^>]*>/gi, "• ")
```
[the remark this answers is not in the document]
Comment (elie222) [reply apps/web/utils/messaging/providers/slack/format.ts]: Addressed in d47298a. This was a valid edge case: adjacent list items were running together because closing `</li>` tags were discarded. The formatter now inserts a newline on `</li>` and the focused Slack formatter test covers that case.

File: apps/web/utils/messaging/providers/slack/format.ts [hunk 3]
```diff
@@ -24,3 +24,25 @@
...
+      .replace(
+        /<a\b[^>]*href=(["'])(.*?)\1[^>]*>(.*?)<\/a>/gis,
+        (_match, _quote, href: string, label: string) => {
+          const normalizedLabel = label.replace(/<[^>]+>/g, "").trim() || href;
```
[the remark this answers is not in the document]
Comment (elie222) [reply apps/web/utils/messaging/providers/slack/format.ts]: Addressed in d47298a. The formatter now tokenizes validated links, escapes stray angle brackets in the remaining text, and adds coverage for unmatched angle brackets so raw tag fragments do not leak through the Slack preview path.

File: apps/web/utils/messaging/providers/slack/format.ts [hunk 4]
```diff
@@ -24,3 +24,25 @@
...
+          return `[${normalizedLabel}](${href})`;
+        },
+      )
+      .replace(/<[^>]+>/g, "")
```
[the remark this answers is not in the document]
Comment (elie222) [reply apps/web/utils/messaging/providers/slack/format.ts]: Addressed in d47298a. The formatter now tokenizes validated links, escapes stray angle brackets in the remaining text, and adds coverage for unmatched angle brackets so raw tag fragments do not leak through the Slack preview path.

File: apps/web/utils/messaging/providers/slack/format.ts [hunk 5]
```diff
@@ -24,3 +24,67 @@
...
+        const token = `SLACK_LINK_TOKEN_${links.length}`;
+        links.push(`<${safeHref}|${safeLabel || safeHref}>`);
+        return token;
+      },
+    )
+    .replace(/<br\s*\/?>/gi, "\n")
+    .replace(/<\/li>/gi, "\n")
+    .replace(/<\/(p|div|blockquote|section|article)>/gi, "\n")
+    .replace(/<li\b[^>]*>/gi, "• ")
+    .replace(/<[^>\n]+>/g, "")
```
[the remark this answers is not in the document]
Comment (elie222) [reply apps/web/utils/messaging/providers/slack/format.ts]: Addressed in a75ad63a. The formatter now uses the same explicit normalizer for list and block tags instead of chaining regex-based tag stripping, which removes the partial-sanitization path the scanner was still flagging here.

File: apps/web/utils/messaging/providers/slack/format.ts [hunk 6]
```diff
@@ -24,3 +24,67 @@
...
```
[the remark this answers is not in the document]
Comment (elie222) [reply apps/web/utils/messaging/providers/slack/format.ts]: Addressed in a75ad63a. I replaced the regex-based tag stripping with a single-pass normalizer that handles the supported HTML subset explicitly and escapes any remaining literal angle brackets as text, so this path no longer relies on partial regex sanitization.

File: apps/web/utils/messaging/providers/slack/format.ts [hunk 7]
```diff
@@ -24,3 +24,67 @@
...
+
+  return links.reduce(
+    (result, link, index) =>
+      result.replaceAll(`SLACK_LINK_TOKEN_${index}`, link),
+    mrkdwn,
+  );
+}
+
+function stripHtmlTags(text: string): string {
+  return text.replace(/<[^>\n]+>/g, "");
```
[the remark this answers is not in the document]
Comment (elie222) [reply apps/web/utils/messaging/providers/slack/format.ts]: Addressed in a75ad63a. The tag-stripping helper was replaced with a single-pass normalizer that escapes literal angle brackets while handling the supported HTML tags explicitly, so this code path no longer leaves raw tag fragments for later processing.

File: apps/web/utils/messaging/providers/slack/format.ts [hunk 8]
```diff
@@ -24,3 +24,106 @@ export function markdownToSlackMrkdwn(text: string): string {
...
+      },
+    ),
+  ).trim();
+
+  const mrkdwn = markdownToSlackMrkdwn(normalized);
+
+  return links.reduce(
+    (result, link, index) =>
+      result.replaceAll(`SLACK_LINK_TOKEN_${index}`, link),
+    mrkdwn,
```
[the remark this answers is not in the document]
Comment (elie222) [reply apps/web/utils/messaging/providers/slack/format.ts:59]: Addressed in ef375179c. Link placeholders now store the exact token we inserted, and token generation skips any value already present in the source text, so user content can no longer collide with the replacement pass.

File: apps/web/utils/messaging/providers/slack/format.ts [hunk 9]
```diff
@@ -24,3 +24,106 @@ export function markdownToSlackMrkdwn(text: string): string {
...
+ */
+export function richTextToSlackMrkdwn(text: string): string {
+  const links: string[] = [];
+
+  const normalized = normalizeSlackRichText(
+    text.replace(
+      /<a\b[^>]*href=(["'])(.*?)\1[^>]*>([\s\S]*?)<\/a>/gi,
+      (_match, _quote, href: string, label: string) => {
+        const safeHref = sanitizeSlackLinkHref(href);
+        const safeLabel = normalizeSlackRichText(label).trim() || href;
```
[the remark this answers is not in the document]
Comment (elie222) [reply apps/web/utils/messaging/providers/slack/format.ts]: Addressed in ef375179c. The empty-label fallback now runs through `normalizeSlackRichText(href)` before building the Slack link label, so special characters are escaped on that path as well.

File: apps/web/utils/messaging/rule-notifications.ts [hunk 10]
```diff
@@ -873,11 +896,37 @@ function buildTerminalCard({
...
+function buildNotificationCardBody(content: NotificationContent): CardChild[] {
+  const children: CardChild[] = [CardText(content.summary)];
+
+  if (content.preview) {
+    children.push(CardText(content.preview));
+  }
+
+  return children;
+}
+
```
[the remark this answers is not in the document]
Comment (elie222) [reply apps/web/utils/messaging/rule-notifications.ts]: Addressed in ef375179c. I moved `buildNotificationCardBody` down with the other private helpers so the exported card builders stay grouped together and the file matches the repo helper-ordering convention.

File: apps/web/utils/messaging/providers/slack/format.ts [hunk 11]
```diff
@@ -24,3 +24,106 @@ export function markdownToSlackMrkdwn(text: string): string {
...
+
+  for (let index = 0; index < text.length; index += 1) {
+    const char = text[index];
+
+    if (char !== "<") {
+      result += escapeSlackTextCharacter(char);
+      continue;
+    }
+
+    const closingIndex = text.indexOf(">", index + 1);
```
[the remark this answers is not in the document]
Comment (elie222) [reply apps/web/utils/messaging/providers/slack/format.ts:83]: Addressed in ef375179c. `normalizeSlackRichText` now preserves existing HTML entities instead of double-escaping them, and unknown `<...>` segments are emitted back as escaped text rather than being dropped.
